### ISSUE TITLE
Not convergence at WFLW dataset

### ISSUE THREAD

**Issue by PlumedSerpent** [HRNet/HRNet-Facial-Landmark-Detection#12]
Hi, thanks for your excellent work. When I  tried to train from pretrained imagenet model on WFLW, it's wired that the loss is convergence at about 0.0011 no matter how I change the lr or the optimizer.BTW,the nme on test set is about 0.20.
All my configuration is just the config file you provided. Do you have any idea about this?

**Comment by flynnamy** [comment]
@PlumedSerpent Have you solved this problem, I have almost same problem like yours.The nme on test set is about 0.12.

**Comment by PlumedSerpent** [comment]
@flynnamy Still no. I turned into 300W dataset, and it worked very well. Maybe I'll try again on WFLW later. Waiting for the author's reply.

**Comment by flynnamy** [comment]
@PlumedSerpent I  get 008 and 010 from 1 to about 0.5 when training, but author get 0.125 on first epoch.How about you when you training?Any suggestion?

**Comment by PlumedSerpent** [comment]
@flynnamy Latest progress--just change the flip option into false, I can get 5.x NME( which means 0.05x), but still, there is a gap away from the SOTA or value in the paper.

**Comment by wondervictor** [comment]
@PlumedSerpent @flynnamy  Code of previous versions is stable and accurate. You can use [this one](https://github.com/HRNet/HRNet-Facial-Landmark-Detection/tree/4c2e5f0cf75a6221fe58947886daeedcdfa3426c) to obtain the SOTA

**Comment by yangyangkiki** [comment]
Hi, our code has been updated, and you can try it on WFLW dataset now.

**Comment by flynnamy** [comment]
Now it seems right.I am trying to train.

**Comment by flynnamy** [comment]
Thanks @yangyangkiki @wondervictor

**Comment by YuxuejieYOKI** [comment]
I have the similar problem on 300W dataset when using pretrained weights, for which the NME is around 1. I just can't figure it out.
![Uploading image.png…]()

**Comment by zhao181** [comment]
> @flynnamy Still no. I turned into 300W dataset, and it worked very well. Maybe I'll try again on WFLW later. Waiting for the author's reply.


How is your result on 300W? did you utilize a pretrained model？ My result is bad

**Comment by PlumedSerpent** [comment]
> > @flynnamy Still no. I turned into 300W dataset, and it worked very well. Maybe I'll try again on WFLW later. Waiting for the author's reply.
> 
> How is your result on 300W? did you utilize a pretrained model？ My result is bad

My result is very close to the paper on 300W, no matter whether I use pretrained model. In detail ,I got 3.34 NME.

**Comment by zhao181** [comment]
Have you notice that the "nms" here means "normalized by ocular distance" ,rather than "normalized by pupil distance"? is this a good result?


On 9/8/2019 00:53，PlumedSerpent<notifications@github.com> wrote：

@flynnamy Still no. I turned into 300W dataset, and it worked very well. Maybe I'll try again on WFLW later. Waiting for the author's reply.

How is your result on 300W? did you utilize a pretrained model？ My result is bad

My result is very close to the paper on 300W, no matter whether I use pretrained model. In detail ,I got 3.34 NME.

**Comment by zhao181** [comment]
> Have you notice that the "nms" here means "normalized by ocular distance" ,rather than "normalized by pupil distance"? is this a good result? On 9/8/2019 00:53，PlumedSerpent<notifications@github.com> wrote： @flynnamy Still no. I turned into 300W dataset, and it worked very well. Maybe I'll try again on WFLW later. Waiting for the author's reply. How is your result on 300W? did you utilize a pretrained model？ My result is bad My result is very close to the paper on 300W, no matter whether I use pretrained model. In detail ,I got 3.34 NME. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.

Is your 3.34 NME means "normalized by ocular distance", or  "normalized by pupil-distance"? I got the result similar to 3.34， but it is normalized by ocular distance. with pupil-distance, the NME is about 0.0489.

**Comment by PlumedSerpent** [comment]
@zhao181  I believe it's a good result,cuz the SOTA before this work is 3.49(also normalized in ocular),which is showed in the paper " Look at Boundary".

**Comment by zhao181** [comment]
In the wingloss paper, the pupil-normalized nms is about 4.04. The "look at boundary" paper uses pupil-normalized nms. I think the hrnet is too complex to get well finetuned

**Comment by zhao181** [comment]
> @zhao181 I believe it's a good result,cuz the SOTA before this work is 3.49(also normalized in ocular),which is showed in the paper " Look at Boundary".

In the wingloss paper, the pupil-normalized nms is about 4.04. The "look at boundary" paper uses pupil-normalized nms. I think the hrnet is too complex to get well finetuned

**Comment by zhao181** [comment]
> @zhao181 I believe it's a good result,cuz the SOTA before this work is 3.49(also normalized in ocular),which is showed in the paper " Look at Boundary".

Thanks. I see it.

**Comment by zhao181** [comment]
> Now it seems right.I am trying to train.

How can we understand the post-processing part in the decode_preds function？

# pose-processing
for n in range(coords.size(0)):
    for p in range(coords.size(1)):
        hm = output[n][p]
        px = int(math.floor(coords[n][p][0]))
        py = int(math.floor(coords[n][p][1]))
        if (px > 1) and (px < res[0]) and (py > 1) and (py < res[1]):
            diff = torch.Tensor([hm[py - 1][px] - hm[py - 1][px - 2], hm[py][px - 1]-hm[py - 2][px - 1]])
            coords[n][p] += diff.sign() * .25
coords += 0.5
preds = coords.clone()

**Comment by oftenliu** [comment]
Hi, thanks for your excellent work. When I tried to train from pretrained imagenet model on WFLW, ,the nme on test set is about 1.X. and the train loss always 0.0017.
All my configuration is just the config file you provided. Do you have any idea about this?

**Comment by yangyangkiki** [comment]
> Hi, thanks for your excellent work. When I tried to train from pretrained imagenet model on WFLW, ,the nme on test set is about 1.X. and the train loss always 0.0017.
> All my configuration is just the config file you provided. Do you have any idea about this?

 Hi, you can try our latest code. we fixed the bug before. Also you can check the code of flip transformation operation.

**Comment by shuangshuangguo** [comment]
I achieve the same result as in paper by using pytorch-1.0, but not by pytorch-0.4.1.
